File: C2/PDOO/Practicas/DeepSpaceJava/src/deepspace/NumericDamage.java

package deepspace;
import java.util.ArrayList;
/**
 *
 * @author Pedro Pablo Ruiz Huertas y Juan Antonio Villegas Recio
 */
public class NumericDamage extends Damage{
    private int nWeapons;
    
    NumericDamage(int w, int s){
        super(s);
        this.nWeapons=w;
    }
    
    NumericDamage(NumericDamage d){
        super(d);
        this.nWeapons=d.nWeapons;
    }
    
    public int getNWeapons(){
        return this.nWeapons;
    }
    
    @Override
    public NumericDamageToUI getUIversion(){
        return new NumericDamageToUI(this);
    }
    
    @Override
    public NumericDamage adjust(ArrayList<Weapon> w, ArrayList<ShieldBooster> s){
        int minShields=nShields;
        if (nShields>s.size())
            minShields=s.size();
        int minWeapons=nWeapons;
            if (nWeapons>w.size())
                minWeapons=w.size();
            
        return new NumericDamage(minWeapons, minShields);
    }
    
    @Override
    public void discardWeapon(Weapon w){
        if(this.nWeapons>0)
            this.nWeapons--;
    }
    
    @Override
    public boolean hasNoEffect(){
        return nShields==0 && nWeapons==0;
    }
    
    @Override
    public String toString(){
        String mensaje="El numero de escudos eliminados ha sido "+nShields+
                        " \ny el numero de armas eliminadas ha sido "+nWeapons;
        return mensaje;
    }
    
    
}

File: C2/PDOO/Practicas/DeepSpaceRuby/lib/NumericDamage.rb

#@author Pedro Pablo Ruiz Huertas y Juan Antonio Villegas Recio

#encoding:utf-8
require_relative "Damage"
require_relative "NumericDamageToUI"

module Deepspace
  class NumericDamage < Damage

    public_class_method :new
    
    attr_reader :nWeapons
    
    def initialize(w,s)
      super(s)
      @nWeapons=w
    end
    
   @override
   def adjust(w,s)
     return NumericDamage.new([@nWeapons,w.length].min, adjustShieldBoosters(s))
   end
    
   def copy
     return NumericDamage.new(@nWeapons,@nShields)
   end
   
   @override
   def discardWeapon(w)
     if @nWeapons>0
      @nWeapons-=1 
     end
   end
   
   @override
   def hasNoEffect
      if @nWeapons==0
        return super
      end
      return false
   end
   
   def getUIversion
     return NumericDamageToUI.new(self)
   end
   
   @override
   def to_s
     return "Dano numerico: #{@nShields} escudos y #{@nWeapons} armas"
   end
   
    
  end
end
